src/ripple/basics/impl/PerfLogImp.cpp

```diff
@@ -79,89 +80,82 @@ PerfLogImp::Counters::countersJson() const
     Rpc totalRpc;
     for (auto const& proc : rpc_)
     {
-        Json::Value p(Json::objectValue);
+        Rpc const* value;
```

Comment:
Since we're no longer making a copy, I think the new code now access the member vars outside the lock.

One possibility is to make the copy before checking the `continue` condition. This is less efficient, but the it isn't _that_ bad (sometimes unnecessarily copying 32 bytes) and the code works out nicely. Of course, we could go back the lambda/optional solution too. I probably prefer that, but I'm OK with either.

Reply:
@mtrippled The three options I see are:
1. Make `value` and optional, check for continue, assign into the optional. This does the minimal amount of copies.
2. Default construct `value`, check for continue, assign into the value. This does an extra default construction.
3. Unconditionally copy the value (under a lock), then check for continue. This does an extra copy sometimes.

I wouldn't expect copying the values to be much more expensive than default constructing - zero out 32 bytes vs. copy 32 bytes. And I think the code is a little nicer - make the variable `const` and get rid of the `continue` check that interrupts the initialization. I prefer (2) myself, but all of these solutions are fine.

I had something like this in mind:
```
Rpc const value = [&] {
            std::lock_guard lock(proc.second.mutex);
            return proc.second.value;
        }();

        if (!value.started && !value.finished && !value.errored)
        {
            continue;
        }
```

I'll also trot out my one line copy under a lock troll-code:  
```
Rpc const value = (std::lock_guard{proc.second.mutex}, proc.second.value);
```